rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,630 @@
+existing `core::sync::atomic::AtomicXyz` API, which expose volatile loads and
+stores of natively supported width with both atomic and non-atomic semantics.
+Deprecate `ptr::[read|write]_volatile` and the corresponding methods of pointer
+types, and recommend that they be replaced with `Relaxed` atomic volatile
+operations on every platform that has support for them.
+
+
+# Motivation
+[motivation]: #motivation
+
+Volatile operations are meant to be an escape hatch that allows a Rust
+programmer to invoke hardware-level memory access semantics that are not
+properly accounted for by the Rust Abstract Machine. They work by triggering the
+(mostly) unoptimized generation of a matching stream of hardware load and store
+instructions in the output machine code.
+
+Unfortunately, the volatile operations that are currently exposed by Rust, which
+map into LLVM's non-atomic volatile operations, unnecessarily differ from the
+memory access semantics of mainstream hardware in two major ways:
+
+1. Concurrent use of volatile memory operations on a given memory location is
+   considered to be Undefined Behavior, and may therefore result in unintended
+   compilation output if detected by the optimizer.
+2. Using an overly wide volatile load or store operation which cannot be carried
+   out by a single hardware load and store instruction will not result in a
+   compilation error, but in the silent emission of multiple hardware load or
+   store instructions, which might be a logic error in the users' program.
+
+By implementing support for LLVM's atomic volatile operations, and encouraging
+their use on every hardware that supports them, we eliminate these divergences
+from hardware behavior and therefore bring volatile operations closer to the
+"defer to hardware memory model" semantics that programmers expect them to have.
+This reduces the odd of mistake in programs operating outside of the regular
+Rust Abstract Machine semantics, which are notoriously hard to get right.
+
+As an unexpected but attractive side-effect, atomic volatile memory operations
+also enable higher-performance interprocess communication between mutually
+trusting Rust programs, through lock-free synchronization of shared memory.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Rust compiler generally assumes that the program that it is building is
+living in a fully isolated memory space. It leverages this knowledge to
+transform said program's memory access patterns for performance optimization
+purposes, under the assumption that said transformations will not have any
+externally observable effect other than the program running faster.
+
+Examples of such transformations include:
+
+- Caching data from RAM into CPU registers.
+- Eliminating unused loads and unobservable stores.
+- Merging neighboring stores with each other.
+- Only updating the part of an over-written struct that has actually changed.
```

Comment:
One important assuming that the compiler makes is that there are no data races. This allows two important transformation:

- When faced with register pressure, a compiler can discard a value that was loaded from memory since it knows that it can read the same value back later if needed. This is invalid if the memory is being concurrently modified.
- Similarly, if the compiler has stored a value to memory then it "knows" that it can later retrieve the same value from that address. Again, this is invalid if the memory is being concurrently modified.

Neither of these limitations apply with volatile operations, since the compiler is not allowed to make any assumptions on the contents of the memory access with volatile operations.

Reply:
IMO, we do not need to introduce the notion of a data race to explain this sort of optimizations. It's just a case of the compiler assuming that the program lives in a fully controlled and isolated address space, where the contents of memory can only change if some action is taken to this end.

I tried to stress this point a bit more in the text.